jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java

```diff
@@ -157,12 +159,60 @@ private void copyResourceToReportDirectory(Resource resource) {
     private void copyResourceToReportDirectory(Resource resource, String targetFilePath) {
         File targetFile = new File(targetFilePath);
         try {
-            FileUtils.copyInputStreamToFile(resource.getInputStream(), targetFile);
+            String resourcePath = String.valueOf(resource.getURI());
+            String jarPath = resourcePath.replace(("!/" + resource.getFilename()), "").replace("jar:file:", "");
+            File jarFile = new File(jarPath);
+
+            if (jarFile.isFile()) {
+                copyResourceFromJar(resource, targetFilePath, targetFile, jarFile);
+            } else {
+                copyResourceFromFile(resource, targetFile);
+            }
         } catch (IOException e) {
             log.error("Cannot copy resource file to target", e);
         }
     }
 
+    private void copyResourceFromFile(Resource resource, File targetFile) {
+        try {
+            if (resource.isFile() && resource.getFile().isDirectory()) {
+                FileUtils.copyDirectory(resource.getFile(), targetFile);
+            } else {
+                FileUtils.copyInputStreamToFile(resource.getInputStream(), targetFile);
+            }
+        }catch (IOException e){
+            log.error("{} is not a file!", resource.getFilename());
+        }
+    }
+
+    private void copyResourceFromJar(Resource resource, String targetFilePath, File targetFile, File jarFile) {
+    try{
+        try (JarFile jar = new JarFile(jarFile)) {
+            JarEntry entry = jar.getJarEntry(resource.getFilename());
+            if (entry.isDirectory()) {
+                jar.stream()
+                    .filter(file -> file.getName().startsWith(entry.getName()) && file.getName().endsWith(".xslt"))
+                    .forEach(file -> {
+                        String fileName = file.getName().replace(entry.getName(), "");
+                        File newFile = new File(targetFilePath + File.separator + fileName);
+                        try {
+                            FileUtils.copyInputStreamToFile(jar.getInputStream(file), newFile);
+                        } catch (IOException e) {
+                            log.error("Unable to get input stream for this file {}", targetFilePath + File.separator + fileName);
```

Comment:
just log IO exception? Shouldn't we re-throw unchecked?

Reply:
Fixed it (I hope :-) )

Reply:
Well yea... It is always good to have it logged, exception may be not obvious enough to some users :smile: 
What about lines `172`, `184` and  `212`? Seems same to me.

Reply:
Line `172` is thrown inside the `log.error` on 173. We did it like that with @paveljandejsek. However I am not entirely sure, what's the difference between doing it this way, and the way I dealt with this one (`throw new UncheckedIOException(e);`)

Reply:
Not thrown, just logged with stacktrace.... So the flow will continue if exception is thrown, we catch it and dont rethrow it. I would probably add the rethrow to those lines as well, like Amy said.

Reply:
So when we log it, it is just written out, but when we throw it, it stops the program?

Reply:
Maybe even better:
`copyResourceFromFile(Resource resource, File targetFile) throws IoException` as it is called from try block above. And same with `copyResourceFromJar(...)`
We can have single place handling same issue.

Reply:
Like this? Seems like an elegant solution.
```suggestion
                                      if (jarFile.isFile()) {
                copyResourceFromJar(resource, targetFilePath, targetFile, jarFile);
            } else {
                copyResourceFromFile(resource, targetFile);
            }
        } catch (IOException e) {
            log.error("Cannot copy resource file to target");
            throw new UncheckedIOException(e);
        }
    }

    private void copyResourceFromFile(Resource resource, File targetFile) throws IOException{
            if (resource.isFile() && resource.getFile().isDirectory()) {
                FileUtils.copyDirectory(resource.getFile(), targetFile);
            } else {
                FileUtils.copyInputStreamToFile(resource.getInputStream(), targetFile);
            }
    }

    private void copyResourceFromJar(Resource resource, String targetFilePath, File targetFile, File jarFile) throws IOException{
            try (JarFile jar = new JarFile(jarFile)) {
                JarEntry entry = jar.getJarEntry(resource.getFilename());
                if (entry.isDirectory()) {
                    jar.stream()
                        .filter(file -> file.getName().startsWith(entry.getName()) && file.getName().endsWith(".xslt"))
                        .forEach(file -> {
                            String fileName = file.getName().replace(entry.getName(), "");
                            File newFile = new File(targetFilePath + File.separator + fileName);
                            try {
                                FileUtils.copyInputStreamToFile(jar.getInputStream(file), newFile);
                            } catch (IOException e) {
                                log.error("Unable to get input stream for this file {}", targetFilePath + File.separator + fileName);
                                throw new UncheckedIOException(e);
                            }
                        });
                } else {
                    log.warn("File name: {}", entry.getName());

                    FileUtils.copyInputStreamToFile(jar.getInputStream(entry), targetFile);
                }
            }
    }
```

Reply:
Like this. Maybe without fancy logging when throwing, but on the other hand little hipster parts makes code juicy :wink:

Reply:
The log is just high-level overview of what's wrong. It is there to help with the first step of finding the root cause of the problem, but it is redundant.